---
title: Working with range data
author: Victor Chen, Bin He
date: 2020-04-09
output: 
  html_document:
    highlight: tango
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Working with Range Data

This workshop is based upon Chapter 9 of Bioinformatic Data Skills by Vincent Buffalo. For the Tuesday portion, you don't need any additional file other than this Rmarkdown itself. To work on it, decide where you will open RStudio -- it is available both on FastX environment and can be easily installed on your local computer (Mac/Windows/Linux). Once you open RStudio, open this Rmarkdown file -- you can get it by either cloning the [2020-Data-Skills](https://github.com/hezhaobin/2020-Data-Skills.git) repository or just copy paste the entire file into a new Rmarkdown file in RStudio.

1. If you have the `bds-files` github repo cloned onto your local computer or on FastX, copy this Rmd file to your `bds-files/chapter-09-working-with-range-data/` and open it there in RStudio.

## Set up 
```{r install_packages, eval=FALSE}
# first examine the R version you have
print(R.version)
# if you have R version 3.6 or later, use the following code
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GenomicRanges")
# if you have earlier versions of R, use the code below 
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
BiocManager::available()
biocLite("GenomicRanges")
```

```{r load_library}
suppressPackageStartupMessages(library(GenomicRanges))
library(ggplot2)
```

To be able to visualize some of the range operations, we will use the custom function written by Vince Buffalo. It is originally stored in an R script file named `plot-ranges.R`. To make this Rmarkdown file self-sufficient, I have included the function in this file.

```{r plot_ranges_function, echo=FALSE}
# plot-ranges.R -- a function for plotting IRanges data
# Copyright (C) 2014 Vince Buffalo <vsbuffaloAAAAA@gmail.com>
# Distributed under terms of the BSD license.

covdf <- function(cov) {
	# from example in IRanges vignette
	cov <- as.vector(cov)
	mat <- cbind(seq_along(cov)-0.5, cov)
	d <- diff(cov) != 0
	mat <- rbind(cbind(mat[d,1]+1, mat[d,2]), mat)
	mat <- mat[order(mat[,1]),]
	data.frame(pos=mat[, 1], cov=mat[, 2])
}

plotIRanges <- function(..., sep=0.5, height=1, set_breaks=TRUE, labcol="grey",
												names=NULL, cov=FALSE, clear=FALSE, disjoint=NULL,
												color=NULL) {
	if (!is.null(colors)) stopifnot(length(colors) <= 3L)
	COLORS <- c("white", "#383838", "#DDDDDD")
	x <- list(...)
	if (!is.null(names))
		names(x) <- names
	dl <- lapply(x, function(d) {
							 out <- as.data.frame(d)
							 out$y <- disjointBins(d)
							 out
							})
	d <- do.call(rbind, dl)
	if (!is.null(disjoint))
		# manually assigned bins
		d$y <- disjoint
	d$ymin <- d$y * (sep + height) - height
	d$ymax <- d$ymin + height
	if (!is.null(color))
		d$color <- color
	if (length(x) > 1 && is.null(names(x)))
		stop("multiple ranges must be given names like plotRanges(rng1=y, rng2=x)")
	if (length(x) > 1)
		d$range <- factor(rep(names(x), sapply(x, length)), names(x))
	p <- ggplot(d)
	if (clear)
		p <- p + geom_rect(aes(ymin=ymin, ymax=ymax, xmin=start-0.5, xmax=end+0.5),
											 fill="white", color="grey30", size=0.3)
	else if(is.null(color))
		p <- p + geom_rect(aes(ymin=ymin, ymax=ymax, xmin=start-0.5, xmax=end+0.5))
	else {
		p <- p + geom_rect(aes(ymin=ymin, ymax=ymax, xmin=start-0.5,
													 xmax=end+0.5, fill=color), color="grey30", size=0.3)
		p <- p + scale_fill_manual("", guide=FALSE,
															 values=COLORS[1:length(unique(color))])
	}
	p <- p + theme_bw()
	if (!is.null(d$names)) {
		p <- p + geom_text(aes(x=start + width/2 - 0.5,
													 y=ymin+(ymax-ymin)/2, label=names), size=8, color=labcol)
	}
	xmin <- min(d$start)
	xmax <- max(d$end)
	xbreaks <- seq(xmin - 1L, xmax + 1L)
	if (set_breaks)
		p <- p + scale_x_continuous(breaks=xbreaks)
	p <- p + theme(panel.grid.major=element_blank(),
								 panel.grid.minor.y=element_blank(),
								 axis.ticks=element_blank())
	if (!cov)
		p <- p + theme(axis.text.y=element_blank())
	p <- p + xlab("") + ylab("")
	if (length(unique(d$range)) > 1)
		p <- p + facet_wrap(~ range, ncol=1)
	if (cov)
		p <- p + geom_line(aes(x=pos, y=cov), covdf(coverage(rngs)), color="red", size=3)
	p
}
```

## A simple example

Let's look at a simple GRanges object and see how it can be useful for genomics analyses
```{r granges_example}
gr <- GRanges(
    seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr
```
Notice a few things:
1. The print out format is very similar to a tibble, that is, it lists the column class beneath the column names, e.g. <Rle> for seqnames and <IRanges> for ranges.
1. Both <Rle> and <IRanges> are new types we have not encountered before (recall <Integer>, <String>, <Factor> and <Boolean> etc.)
    - keep both types in your mind. we will study them in more details in this workshop.

Let's look at how we can access the information in a granges object, that is, how we can "retrieve" information.
```{r access_granges_object}
seqnames(gr) # extract just the sequence names
ranges(gr)   # extrant just the ranges
strand(gr)   # extract just the strand information
granges(gr)  # extract the genomic ranges without the metadata (score and GC)
mcols(gr)    # extract just the metadata
mcols(gr)$score # note that mcols(gr) returns a data.frame

# Information about the lengths of the various sequences that the ranges are aligned to can also be stored in the GRanges object. So if this is data from Homo sapiens, we can set the values as:
seqlengths(gr) <- c(249250621, 243199373, 198022430)
# And then retrieve as
seqlengths(gr)
```

Next we will take a closer look at one of the key abstractions (elements) in a GRanges object, the ranges.

## IRanges
We will familiarize ourselves with the concept of ranges by practicing with generic range objects using the IRanges package. GenomicRanges is a specialized version of IRanges.

### Create and access IRanges objects

Let's look at how we can create an IRanges object and how to access its various parts.

```{r explore_IRanges}
# load the IRanges package (packages extend the functions of the base R)
library(IRanges) # you might see some package startup messages when you run this
library(ggplot2) # the custom plotIRanges() function uses ggplot2

# 1. Create IRanges objects
#    to construct an IRanges, you need to use the constructor function IRanges()
rng <- IRanges(start=4, end=13)
print(rng) # this shows the content of rng. notice how the output is similar/different from a data frame or a tibble
class(rng) # show the class of the object
str(rng)   # look inside an IRanges object, with 6 "slots"

# 2. IRanges with specific width
#    Instead of specifying the start and end position, one can replace end with the width of the range
IRanges(start=4, width=3)
IRanges(end=5, width=5)

# 3. Create multiple ranges in one-go
x <- IRanges(start=c(4, 7, 2, 20), end=c(13, 7, 5, 23))
print(x)

# 4. You can name each range by doing the following
names(x) <- letters[1:4] # name the range in letter
print(x)

# 5. Accessor functions
#    Remember the different "slots" we see with str(rng)? We can "access" them using special accessor functions
start(x)
end(x)
width(x)
end(x) <- end(x)+4 
print(x)

# 6. Subset IRanges
#    Even though the IRanges object looks like a "table", it's internally stored as a list of individual ranges (i.e. records), and you can subset them just likely you can with a vector
x[2:3] # The 2 and 3 row of x
x[start(x)<5] # This will show the subset of ranges with start positions smaller than 5
x[width(x)>8] # This will show the subset of ranges with start positions bigger than 8
x['a']        # The row with name "a"

# 7. Concatenate IRanges -- c() works for this!
a <- IRanges(start = 7, width=4)
b <- IRanges(start = 2, end = 5)
c(a,b) # Using the function c() to merge ranges
```

### Basic Range Operations: Arithmetic, Transformations, and Set Operations

Often times we would like to manipulate ranges, such as enlarging or shrinking them, or performing common "set operations" such as union, difference, etc. With IRanges, you can do these intuitively and fast.

#### Operations on single ranges
```{r basic_operations_on_single_ranges}
# 1. Expand or contract (symmetrically)
x <- IRanges(start=c(40, 80),end = c(67,114))
x+4L
x-10L
plotIRanges(x=x, "x, end position extended by 4"=x+4L, "x, end position reduced by 10"=x-10L, set_breaks = FALSE)

# 2. How to cut off the head/tail of ranges based on a set boundary?
y <- IRanges(start = c(4, 6, 10, 12), width=13)
y
restrict(y, 5, 10) # cuts a set of ranges such that they fall inside of certain bound
plotIRanges(y=y, "y, restricted to (5,10)"=restrict(y, 5, 10))

flank(x, width = 7) # flank creates ranges width positions upstream of the ranges 
flank(x, width = 7, start=FALSE) # By setting start=false, it flanking ranges downstream
plotIRanges(x=x, "flank(x, width=7)"=flank(x,width=7), "flank(x, width=7, start=FALSE)"=flank(x, width=7, start=FALSE), set_breaks = FALSE)
```

#### Working on multiple ranges

```{r basic_operations_on_multiple_ranges}
# 1. Reduce
set.seed(0) # this sets the seed for the pseudo-random number generator
alns <- IRanges(start = sample(seq_len(50),20), width = 5) 
# seq_len(50) = 1:50
# sample(seq_len(50), 20) randomly samples 20 values from 1:50 without replacement
head(alns, 4) 
reduce(alns) # reduce takes a set of possibly overlapping ranges and reduces them to a set of nonoverlapping ranges

# 4. Gaps
gaps(alns) #find the gap between the ranges
plotIRanges(alns=alns, "alns, reduced"=reduce(alns), "alns, gaps"=gaps(alns), set_breaks = FALSE)
```

#### Set operations

Difference, intersection, union
```{r set_operations}
a <- IRanges(start=4, end=13)
b <- IRanges(start=12, end=17)
intersect(a, b)
setdiff(a, b)
setdiff(b, a)
union(b, a)
union(a, b)
plotIRanges("a,b"=c(a,b), "intersect(a,b)"=intersect(a, b), "setdiff(a,b)"=setdiff(a, b), "setdiff(b,a)"=setdiff(b,a), "union(a,b)"=union(b, a))
```

### FindOverlaps
Finding overlaps is a very common operation required in genomics, such as counting reads based on gene annotations. It is seemingly simple but in reality quite complicated, not because of the computation being hard, but because of the intricacy of defining "overlap". Here we will learn the basics of finding overlaps. Later we will learn more specific operations on genomic ranges

```{r find_overlaps}
# this operation is direction, that is we start with a "query" and look for overlaps in a "subject"
qry<- IRanges(start = c(1, 26, 19, 11, 21, 7), end = c(16, 30, 19, 15, 24, 8), names = letters[1:6])
sbj<- IRanges(start=c(1, 19, 10), end=c(5, 29, 16), names = letters[24:26])
plotIRanges(query=qry,subject=sbj)
hts<- findOverlaps(qry, sbj) # find the overlap between the qry and sbj
hts

names(qry)[queryHits(hts)]
names(sbj)[subjectHits(hts)] # find the name of each query and subject range with an overlap

hts_within<- findOverlaps(qry, sbj, type = "within") # limit our overlap results to only include query ranges that fall entirely within subject ranges
hts_within

findOverlaps(qry, sbj, select = "first")
findOverlaps(qry, sbj, select = "last")
findOverlaps(qry, sbj, select = "arbitrary")

# accessor functions for the internal components of a "hit" object
as.matrix(hts) # Hits objects can be coerced to matrix
countQueryHits(hts) # How many subject ranges each query IRanges object overlaps
setNames(countQueryHits(hts), names(qry))
countSubjectHits(hts) # How many query ranges overlap the subject ranges
setNames(countSubjectHits(hts), names(sbj))

# SubsetByOverlaps and countOverlaps
countOverlaps(qry, sbj) # countoverlaps is similar to the soultion using countQueryOverlaps and setnames
subsetByOverlaps(qry, sbj) #it is same as qry[unique(queryHits(hts))]
```
### Finding Nearest Ranges and Calculating Distance
This is another common operations on ranges, namely finding the nearest ranges and calculating distance from it. Imagine that you have experimentally determined the locations in the genome bound by some important regulatory proteins. Not knowing which gene these proteins regulate, you best guess is that they regulate the gene that is the closest to the protein (often not a bad assumption to start with). Thus your tasks is, given two set of ranges -- the first for the places where the protein binds and the second the "address" for every gene in the genome -- to find the nearest gene to each binding site. Here we will learn how to do this with the abstraction of IRanges
```{r nearest_ranges_and_distance}
qry<- IRanges(start=6, end = 13, names = 'query')
sbj<- IRanges(start = c(2, 4, 18, 19), end = c(4, 5, 21, 24), names = 1:4)
plotIRanges(query=qry, subject=sbj)

nearest(qry, sbj) # return the nearest range, regardless of whether it's upsteam or downstream of the query 
precede(qry, sbj) # return the nearest range that the query is upstream
follow(qry, sbj) # return the nearest range that the query is downstream

# Distance to Nearest 
qry<- IRanges(sample(seq_len(1000), 5), width = 10)
sbj<- IRanges(sample(seq_len(1000), 5), width = 10)
plotIRanges(query=qry, subject=sbj, set_breaks = FALSE)
distanceToNearest(qry, sbj) # It finds the closest subject range, and returns everything in a Hits object with an additional column indicating the distance
distance(qry, sbj) # returns each pairwise distance between query and subject ranges
```
### Calculating coverages
A common task in genomics analyses is to calculate the coverage, that is, for each position in a genome, how many reads map to that position or range. For example, one may encounter this while sequencing a new genome. Because of the sequencing errors, one usually try to cover each position in that genome with multiple reads, so that sequencing errors can be identified and removed by spotting inconsistence between a subset of the reads at a particular position with the majority of the reads at that position. 

Calculating coverage can be thought of as a special instance of counting overlaps -- when the subject ranges are precisely length of 1 and tiles across the genome (challenge: what would be the mode of counting overlaps in the application to coverage calculation?).

What's special about calculating coverages compared with the type of counting overlaps we just discussed is the data density -- the list of subject ranges is exactly as large as the number of nucleotides (or any other sequences) as the genome (or proteome, etc.). This raises a practical problem: to represent that many ranges and doing calculation over all of them becomes computationally expensive, especially for memory. To solve this problem, researchers have come up with a clever way of encoding coverage data, which is termed run length encoding. We will see what Run Length Encoding is and how it is used in calculating coverages.

#### Run Length Encoding and Views 
```{r run_length_encoding}
x <- as.integer(c(4, 4, 4, 3, 3, 2, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 4, 4, 4, 4, 4, 4, 4))
sprintf("The length of the vector x is %i", length(x))
xrle <- Rle(x) # Run-length encoding compresses this sequence
xrle          # note that instead of 
sprintf("The length of the run length object xrle is %i", length(runValue(xrle)))
# however, in reality the run length object not only store the values and length of each "run", but also other accessory information, making them bigger than the original object when the latter is a small vector. This is often termed memory overheads. What this menas is that Rle is not efficient unless the original vector is rather big. 
str(xrle)
# Let's comapre the memory used by the original and Rle encoded objects.
sprintf("The size of the vector x is %i bytes", object.size(x))
sprintf("The size of the Rle object xrle is %i bytes", object.size(xrle))

as.vector(xrle) # revert back to vector 
```

#### (Optional) Basic operations of Run-length encoding
```{r basic_operations_Rle}
xrle + 4L
xrle  /2
xrle > 3
xrle[xrle > 3]
sum(xrle)
summary(xrle)
round(cos(xrle), 2)

runLength(xrle) # Rle object's lengths 
runValue(xrle) # Rle object's values
```

#### Coverage(where we encounter Rle)
```{r coverage}
set.seed(0)
rngs <- IRanges(start = sample(seq_len(60), 20), width=7)
names(rngs) <- LETTERS[1:20] # label one range for examples later 
plotIRanges(rngs, set_breaks = FALSE)

rngs_cov <- coverage(rngs)
rngs_cov[rngs['A']] # what the coverage was in the region overlapping the range laberled "A"
mean(rngs_cov[rngs["A"]]) # find the mean coverage within this range
plotIRanges(rngs, cov = TRUE, set_breaks = FALSE)

rngs_cov > 3 # where is coverage greater than 3
rngs_cov[as.vector(rngs_cov)>3] # extract the depths that are greater than 3 
```

#### (Optional) Going from run-length encoded sequences to ranges with slice()
Sometimes you want to create ranges based on the coverage data. For example, after mapping your reads to the reference genome, you would like to identify all regions covered by at least two reads, or, conversely, you want to find out all regions with less than two reads, which you may consider leaving out of your analysis. You can do this with the `slice()` function, which operates on an Rle object.
```{r from_Rle_to_ranges}
min_cov2 <- slice(rngs_cov, lower=2) # take our coverage Rle object rngs_cov and slice it to create ranges corresponding to regions with more than 2x coverage
min_cov2
```

#### (Optional) Advanced IRanges:Views
```{r views}
viewMeans(min_cov2) # find the mean
viewMaxs(min_cov2) # find the Maximum 
viewApply(min_cov2, median) # find the median 

length(rngs_cov)
bwidth <- 5L # we will set bwidth to be our bin width
end <- bwidth * floor(length(rngs_cov)/ bwidth) # we compute the end position of our window
windows <- IRanges(start = seq(1, end, bwidth), width = bwidth) # create our windows using IRanges
head(windows)
cov_by_wnd <- Views(rngs_cov, windows) # create our views object
head(cov_by_wnd)
viewMeans(cov_by_wnd) # get the mean coverage per window 
```

**Below are for Thursday, but feel free to plow ahead!**


## Storing genomic ranges with GenomicRanges 
The `GRanges` build off of `IRanges` by adding two other pieces of information necessary to specify a genomic location: sequence name (e.g. which chromosome) and strand (positive/negative, sometimes called Watson vs Crick, after the two Nobel Prize Winnders for DNA structure). `GRanges` objects also have _metadata columns_, which are the data linked to each genomic range. Below we will learn how to construct and manipulate `GRanges`.

```{r install_and_load_packages, eval=FALSE}
# first examine the R version you have
print(R.version)
# if you have R version 3.6 or later, use the following code
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GenomicRanges")
# if you have earlier versions of R, use the code below 
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
BiocManager::available()
biocLite("GenomicRanges")
# load GenomicRanges
suppressMessages(library(GenomicRanges))
library(ggplot2)
```

```{r creating_granges_object}
gr <- GRanges(seqnames = c("chr1", "chr1", "chr2", "chr3"), ranges = IRanges(start = 5:8, width = 10), strand = c("+", "-", "-", "+")) # we can create GRanges objects much like we didi with IRanges objects
gr

gr <- GRanges(seqnames = c("chr1", "chr1", "chr2", "chr3"), ranges = IRanges(start = 5:8, width = 10), strand = c("+", "-", "-", "+"), gc=round(runif(4),3)) # add arbitrary metadata columns by specifying additional named arguments 
gr

# The latest version of GRanges contains a number of useful attributes, such as sequence length and genome name
seqlens <- c(chr1=152, chr2=432, chr3=903)
gr <- GRanges(seqnames = c("chr1", "chr1", "chr2", "chr3"), ranges = IRanges(start = 5:8, width=10), strand = c("+", "-", "-", "+"), gc=round(runif(4),3), seqlengths=seqlens)
seqlengths(gr) <- seqlens# we can specify the sequence lengths in the GRanges constructor, or set it after the object has been created using the seqlengths function 
genome(gr) <- "Alien"           # name of the genome, whatever it maybe!
isCircular(gr) <- rep(FALSE, 3) # if you have a bacterial genome or a plasmid, set this attribute to TRUE
seqinfo(gr)                     # this shows the information associated with the GRanges object
gr 
```

## Accessor functions in GRanges
You can access various information contained in a `GRanges` object using the following accessor functions. Pay special attention to the output -- notice how some of the information are encoded as run length, which we encountered in the Tuesday workshop.

```{r access_information_in_granges}
seqnames(gr) # sequence name 
strand(gr)  # strand
class(strand(gr)) # this shows you that the strand information is encoded as a run length
ranges(gr)  # extract all IRanges ranges from a GRanges object
length(gr)  # length of GRanges
names(gr) <- letters[1:length(gr)]
gr

start(gr) > 7 # if you want all ranges with a start position greater than 7
gr[start(gr) > 7]

table(seqnames(gr))
gr[seqnames(gr)=="chr1"] # count how many ranges there are per chromosome and then subset to include only ranges fro a particular chromosome

mcols(gr) # this is used to access ALL metadata columns. it retunrs a dataframe

mcols(gr)$gc
gr$gc # shortcut for accessing a metadata column 
```
